client/utils/guidance.py

```diff
@@ -0,0 +1,210 @@
+from client.utils import assess_id_util
+import json
+import os
+import sys
+
+"""
+This utility is called by unlock.py. This guidance utility changes the message that students see
+after they input in wrong answers. If a student guesses a certain amount of wrong answers
+that all show a certain type of confusion about the problem, this utility will instead of showing
+the default "Not Quite Try Again" message will show some kind of message that will target 
+that type of misunderstanding. 
+
+This utility object requires internet access to determine what treatmeng group they are assigned
+to. The different treatment groups will have varying threshold level of answers as well as different
+messages and other differences. It will contact the server defined below in the variable TGSERVER with
+the user's email and lab assignment to get the treatment group number. 
+
+Commonly used acronyms: 
+TG = treatment group number
+KI = Type of targeted understanding
+misU = Type of misunderstanding the student is showing
+"""
+
+TGSERVER = "http://tg-server.app.cs61a.org/"
+LOCAL_TG = "tests/tg.ok_tg"
+OK_GUIDANCE = "tests/.ok_guidance"
+DEFAULT_MSG = "-- Not quite. Try again! --"
+DEFAULT_RETURN_VALUE = ({}, -1, "")
+COUNT_FILE_PATH = "tests/misUcount.json"
+TG_SERVER_ENDING = "/unlock_tg"
+TG_CONTROL = 0
+# If student forces guidance messages to show, we will assign treatment group number below
+GUIDANCE_FLAG_TG_NUMBER = 1
+TG_ERROR_VALUE = -1
+
+lambda_string_key_to_func = {
+    'none': lambda info, strMisU: None, 
+    'ki': lambda info, strMisU: info['ki'],
+    'misU2Msg': lambda info, strMisU: info['dictMisU2Msg'].get(strMisU)
+}
+  
+class Guidance():
+    def __init__(self,current_working_dir):
+        """
+        Initializing everything we need to the default values. If we catch
+        an error when opening the JSON file, we flagged it as error.
+        """
+        self.tg_id = -1
+        self.misU_count = {}
+        self.current_working_dir = current_working_dir
+        try:
+            with open(current_working_dir + OK_GUIDANCE,"r") as f:
+                self.guidance_json = json.load(f)
+            self.error = False
+        except:
+            self.error = True
+
+
+    def guidance_msg(self,unique_id, input_lines,access_token,hash_key,guidance_flag):
+        """ 
+        Based on the student's answer (input_lines), we grab each associated
+        message if its corresponding misunderstanding's count is above the threshold
+        """    
+        if self.error:
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        shorten_unique_id = assess_id_util.canon(unique_id)
+        # Try to get th info dictionary for this question
+        wa_2_dict_info = self.guidance_json['dictAssessId2WA2DictInfo'].get(shorten_unique_id)
+
+        if not wa_2_dict_info: # get returns None if wa_2_dict_info doesn't have shorten_unique_id in dictionary
+            print (DEFAULT_MSG)
```

Comment:
Please remove the space between `print` and `(` in all of your print statements.